cl/corpus_importer/tasks.py

```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+    :return: None
+    """
+
+    r = get_redis_interface("CACHE")
+    probe_iteration = 1
+    latest_match = 0
+    time_out_counter = 0
+    consecutive_empty_probes = 0
+    iquery_pacer_case_id_final = int(
+        r.hget("iquery_pacer_case_id_final", court_id) or 0
+    )
+
+    # Probe limit e.g: 9 probe iterations -> 256
+    probe_limit = 2 ** (settings.IQUERY_PROBE_ITERATIONS - 1)
+    court_probe_cycle_no_hits = r.hincrby(
+        "court_probe_cycle_no_hits", court_id, 1
+    )
+    reports_data = []
+    while probe_iteration <= settings.IQUERY_PROBE_ITERATIONS:
+        if time_out_counter >= 3:
+            logger.warning(
+                "Aborting probing task for %s after 3 consecutive iterations "
+                "that timed out.",
+                court_id,
+            )
+            r.setex(
+                f"court_wait:{court_id}",
+                settings.IQUERY_COURT_BLOCKED_WAIT,
+                2,
+            )
+            break
+
+        probe_iteration, pacer_case_id_to_lookup = compute_next_binary_probe(
+            iquery_pacer_case_id_final,
+            probe_iteration,
+            court_probe_cycle_no_hits,
+            probe_limit,
+        )
+        try:
+            report_data = query_iquery_page(court_id, pacer_case_id_to_lookup)
+        except HTTPError:
+            # Set expiration accordingly and value to 2 to difference from
+            # other waiting times.
+            r.setex(
+                f"court_wait:{court_id}",
+                settings.IQUERY_COURT_BLOCKED_WAIT,
+                2,
+            )
+            logger.warning(
+                "HTTPError occurred when crawling iquery. The court %s website "
+                "is probably down or has blocked us. Abort probing for %s seconds ",
+                court_id,
+                settings.IQUERY_COURT_BLOCKED_WAIT,
+            )
+            delete_redis_semaphore("CACHE", make_iquery_probing_key(court_id))
+            return None
+
+        except requests.Timeout as exc:
+            logger.warning(
+                "The court website is probably down. Waiting a bit before trying "
+                "the next pacer_case_id."
+            )
+            time.sleep(settings.IQUERY_COURT_TIMEOUT_WAIT)
+            time_out_counter += 1
+            continue
+
+        if report_data:
+            # Find and update/store the Docket.
+            reports_data.append((pacer_case_id_to_lookup, report_data))
+            latest_match = pacer_case_id_to_lookup
+            # Restart consecutive_empty_probes
+            consecutive_empty_probes = 0
+        else:
+            consecutive_empty_probes += 1
+            if (
+                consecutive_empty_probes
+                >= settings.IQUERY_PROBE_MAX_CONSECUTIVE_FAILURES
+            ):
+                # Too many consecutive blank probes in a row.
+                # Terminate the probing loop.
+                break
```

Comment:
Well, I was using a similar approach, but with the difference that it checked 2 missed probes before aborting.

1: Hit
2: Miss
4: Miss
Abort

The problem here is that this is not enough to pass some small gaps, especially if they are at the beginning of the next probe.

For example, in this case, it seems common for the next pacer case to be blank, causing the probe to be aborted.

It is also possible that 4 is still blank, but 8 can be a hit.

The more probes we perform, the less likely it is that the process will get stuck in the first blank items.

1: Hit
2: Miss
4: Miss
8: Hit

1: Hit
2: Miss
4: Miss
8: Miss
16: Hit

So, setting up a counter for missed hits allows us to pass these kinds of gaps more easily. 
Does it make sense?

Reply:
I think so, but I'm suggesting that until you get a hit, you just go up to 256, but once you get a hit you abort as soon as you get a miss. 

So the algo is:
 - Probe until you get a hit.
 - Once you get a hit, keep probing until you get a miss.

In five minutes, you'll come back and do the same, which will get you past any gaps.

Reply:
> I'm suggesting that until you get a hit, you just go up to 256, but once you get a hit you abort as soon as you get a miss.

Great, just to confirm I understood correctly:

Every probe task will always have to perform at least 9 probes?
`1, 2, 4, 8, 16, 32, 64, 128, 256`

And if 256 is a hit, we should still probe 512, 1024, and so on until we get a miss and then abort?

Because the other interpretation of this algo, which is the problem I was describing, is:

Once you get a hit (let's say 1 is a hit), keep probing until you get a miss (2 is a miss). Probing is aborted on 2.
Which is not good.

Reply:
Not quite. We'll go up to 256 every five minutes. We'll keep going higher if we get misses or hits, but once we get hits, we'll abort on the first subsequent miss.

> Once you get a hit (let's say 1 is a hit), keep probing until you get a miss (2 is a miss). Probing is aborted on 2.

That's OK. We'll get it on the next round.

Reply:
Got it! Sounds good now, I'll apply it.

Reply:
This algo is now implemented:

> So the algo is:
> 
> Probe until you get a hit.
> Once you get a hit, keep probing until you get a miss.

I tested it on a couple of courts and it worked smoothly!